ui-v2/tests/steps.js

```diff
@@ -0,0 +1,232 @@
+        }
+      )
+      // interactions
+      .when('I visit the $name page', function(name) {
+        currentPage = pages[name];
+        return currentPage.visit();
+      })
+      .when('I visit the $name page for the "$id" $model', function(name, id, model) {
+        currentPage = pages[name];
+        return currentPage.visit({
+          [model]: id,
+        });
+      })
+      .when(
+        ['I visit the $name page for yaml\n$yaml', 'I visit the $name page for json\n$json'],
+        function(name, data) {
+          currentPage = pages[name];
+          return pages[name].visit(data);
+        }
+      )
+      .when('I click "$selector"', function(selector) {
+        return click(selector);
+      })
+      .when('I click $prop on the $component', function(prop, component) {
+        // Collection
+        var obj;
+        if (typeof currentPage[component].objectAt === 'function') {
+          obj = currentPage[component].objectAt(0);
+        } else {
+          obj = currentPage[component];
+        }
+        const func = obj[prop].bind(obj);
+        try {
+          return func();
+        } catch (e) {
+          console.error(e);
+          throw new Error(`The '${prop}' property on the '${component}' page object doesn't exist`);
+        }
+      })
+      .when('I submit', function(selector) {
+        return currentPage.submit();
+      })
+      .then('I fill in "$name" with "$value"', function(name, value) {
+        return currentPage.fillIn(name, value);
+      })
+      .then(['I fill in with yaml\n$yaml', 'I fill in with json\n$json'], function(data) {
+        return Object.keys(data).reduce(function(prev, item, i, arr) {
+          return prev.fillIn(item, data[item]);
+        }, currentPage);
+      })
+      .then(['I type with yaml\n$yaml'], function(data) {
+        const keys = Object.keys(data);
+        return keys
+          .reduce(function(prev, item, i, arr) {
+            return prev.fillIn(item, data[item]);
+          }, currentPage)
+          .then(function() {
+            return Promise.all(
+              keys.map(function(item) {
+                return triggerKeyEvent(`[name="${item}"]`, 'keyup', 83);
+              })
+            );
+          });
+      })
+      // debugging helpers
+      .then('print the current url', function(url) {
+        console.log(currentURL());
+        return Promise.resolve();
+      })
+      .then('log the "$text"', function(text) {
+        console.log(text);
+        return Promise.resolve();
+      })
+      .then('pause for $milliseconds', function(milliseconds) {
+        return new Promise(function(resolve) {
+          setTimeout(resolve, milliseconds);
+        });
+      })
+      // assertions
+      .then('a $method request is made to "$url" with the body from yaml\n$yaml', function(
+        method,
+        url,
+        data
+      ) {
+        const request = api.server.history[api.server.history.length - 2];
```

Comment:
This looks brittle. Is it guaranteed that the request in question will always be the second-to-last one made?

Reply:
hehe agree completely here, this is on my todo list. This is specific to my needs right now, and isn't 'brittle' as such, just not as I would like it. It's a sufficient test for what I need it for right now though.